BE: Mozfest Current Events Slider

b-ggs opened the pull request:
Related PRs/issues #7431 

## Implementation Checklist

- [X] Editor can add a heading (required)
- [X] Editor can add items (unlimited)
- [X] Editor is advised that no more than 5 items are added
- [X] Editor can add a heading to an item (required)
- [X] Editor can add a link to an item
- [X] Editor can add a image to an item (required)
- [X] Editor can add a plain text description to an item (required)
- [X] Editor can add 2 buttons with either link to file (ical) external URL or internal page (1 required, 2nd optional)

## Checklist

**Changes in Models:**
- [X] Did I squash my migration?

Here is a comparison of the dummy templates I modified from an existing component against the designs on Figma:

![image](https://user-images.githubusercontent.com/6130147/135029577-dc59ea99-9909-41d8-9011-934754d7df62.png)

SharmaineLim commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/current_events_slider_list_block.py:
```diff
@@ -0,0 +1,73 @@
+from wagtail.images.blocks import ImageChooserBlock
+
+
+class CurrentEventBaseButtonBlock(blocks.StructBlock):
+    label = blocks.CharBlock(help_text='Button label')
+
+
+class CurrentEventButtonInternalPage(CurrentEventBaseButtonBlock):
+    link = blocks.PageChooserBlock(help_text='Page that this button should link out to.')
+
+
+class CurrentEventButtonExternalPage(CurrentEventBaseButtonBlock):
+    link = blocks.URLBlock(help_text='URL that this button should link out to.')
```
These are such generic things they can probably be shared blocks in `utils/blocks` or something like that. I find it highly unlikely these will change and if we make these for every card block we create in MozFest, one day we're just going to have:

- `CurrentEventBaseButtonBlock`, `CurrentEventButtonInternalPage`, `CurrentEventButtonExternalPage`
- `PastEventBaseButtonBlock`, `PastEventButtonInternalPage`, `PastEventButtonExternalPage`
- `FutureEventBaseButtonBlock`, `FutureEventButtonInternalPage`, `FutureEventButtonExternalPage`
- `RelatedEventBaseButtonBlock`, `RelatedEventButtonInternalPage`, `RelatedEventButtonExternalPage`
- `BaseButtonBlock`, `NewsButtonInternalPage`, `NewsButtonExternalPage`
- `RelatedBaseButtonBlock`, `RelatedNewsButtonInternalPage`, `RelatedNewsButtonExternalPage`

which seems a bit excessive and repetitive.

I think we can also add the `link_url` property to `CurrentEventBaseButtonBlock` instead of having a completely separate `CurrentEventBlockStructValue` with `buttons`? Feel free to correct me if this one's not viable in some way.

SharmaineLim commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/current_events_slider_list_block.py:
```diff
@@ -0,0 +1,73 @@
+
+    link_label = blocks.CharBlock('Label of the link below the heading.')
+
+    link_url = blocks.URLBlock('URL of the link below the heading.')
```
Is this always going to link to an external site? Will we ever want to link to an internal page?

b-ggs replied:
I've modified this to be able to link to either an external or internal page

b-ggs commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/common/link_blocks.py:
```diff
@@ -0,0 +1,43 @@
+from wagtail.core import blocks
+
+
+class InternalLinkBlockStructValue(blocks.StructValue):
+    @property
+    def link_url(self):
+        return self.get('link').url
+
+
+class ExternalLinkBlockStructValue(blocks.StructValue):
+    @property
+    def link_url(self):
+        return self.get('link')
+
+
+class InternalLinkBlock(blocks.StructBlock):
+    link = blocks.PageChooserBlock(help_text='Page that this should link out to.')
+
+    class Meta:
+        value_class = InternalLinkBlockStructValue
+
+
+class ExternalLinkBlock(blocks.StructBlock):
+    link = blocks.URLBlock(help_text='URL that this should link out to.')
+
+    class Meta:
+        value_class = ExternalLinkBlockStructValue
+
+
+class LabelledLinkBaseBlock(blocks.StructBlock):
+    label = blocks.CharBlock(help_text='Label for this link.')
+
+
+class LabelledInternalLinkBlock(InternalLinkBlock, LabelledLinkBaseBlock):
+    pass
+
+
+class LabelledExternalLinkBlock(ExternalLinkBlock, LabelledLinkBaseBlock):
+    pass
```
@SharmaineLim I've moved out the reusable blocks to here

stevedya commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/current_events_slider_list_block.py:
```diff
@@ -0,0 +1,39 @@
+
+    body = blocks.TextBlock(help_text='Body text of the card.')
+
+    buttons = blocks.StreamBlock(
```
Suggestion: We may need a document link block option as well here for iCal files

stevedya commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/current_events_slider_list_block.py:
```diff
@@ -0,0 +1,39 @@
+from .common.link_blocks import LabelledExternalLinkBlock, LabelledInternalLinkBlock
+
+
+class CurrentEventBlock(blocks.StructBlock):
```
Could we give this block some meta info so it is labeled in the gui of wagtail 😄

SharmaineLim commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/common/link_blocks.py:
```diff
@@ -0,0 +1,43 @@
+        return self.get('link')
+
+
+class InternalLinkBlock(blocks.StructBlock):
+    link = blocks.PageChooserBlock(help_text='Page that this should link out to.')
+
+    class Meta:
+        value_class = InternalLinkBlockStructValue
+
+
+class ExternalLinkBlock(blocks.StructBlock):
+    link = blocks.URLBlock(help_text='URL that this should link out to.')
+
+    class Meta:
+        value_class = ExternalLinkBlockStructValue
```
Does having this not work?

```python
class InternalLinkBlock(blocks.StructBlock):
    link = blocks.PageChooserBlock(...)

    @property
    def link_url(self):
        return self.get('link')
```

Or is the goal to make `link` completely unavailable outside of these blocks?

b-ggs replied:
This won't work for StructBlocks unfortunately: https://docs.wagtail.io/en/latest/advanced_topics/customisation/streamfield_blocks.html#additional-methods-and-properties-on-structblock-values

Adding properties to a StructBlock requires you to create a StructValue

SharmaineLim commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/current_events_slider_list_block.py:
```diff
@@ -0,0 +1,39 @@
+
+
+class CurrentEventBlock(blocks.StructBlock):
+    title = blocks.CharBlock('Heading for the card.')
```
Should this have `help_text=`?

SharmaineLim reviewed (approved):
Overall, LGTM. Left a couple of questions/potential tweaks.

b-ggs commented:
I've added a new button type for documents, specifically for use with iCal documents:

![image](https://user-images.githubusercontent.com/6130147/136014966-b894f770-3159-4ef5-aff6-66a81ad55af4.png)

Additionally, I've added more help texts for additional clarity.

--- 

@stevedya:
>Could we give this block some meta info so it is labeled in the gui of wagtail 😄

I've added a help text under Current Events here:

![image](https://user-images.githubusercontent.com/6130147/136015294-1c6b05bf-f6c2-4513-8ec5-21daaa21f280.png)

Does that suffice or did you have something else in mind?
